File: services/rust/salvo/src/api/handlers/post.rs
use crate::{api::handlers::state, task::Task};
use salvo::{Depot, Request, http::StatusCode, writing::Json};
use serde_json::Value;
use tracing::{error, info, instrument};
use uuid::Uuid;

#[salvo::handler]
#[instrument(skip_all)]
pub async fn post_handler(
    depot: &mut Depot,
    request: &mut Request,
) -> Result<Json<Value>, StatusCode> {
    let state = state(depot);

    let request = request
        .parse_json::<Task>()
        .await
        .map_err(|_| StatusCode::BAD_REQUEST)?;

    let id = Uuid::new_v4();

    if let Ok(mut state) = state.lock() {
        state.tasks.insert(id, request.clone());
        drop(state);

        info!(
            %id,
            secret = request.secret,
            operation = request.operation.to_string(),
            method = "POST",
            "Inserted new task"
        );

        return Ok(Json(Value::String(id.to_string())));
    }

    error!(
        %id,
        secret = request.secret,
        operation = request.operation.to_string(),
        method = "POST",
        "Poisoned lock"
    );

    Err(StatusCode::INTERNAL_SERVER_ERROR)
}

File: services/go/echo/routes/handlers/post.go
package handlers

import (
	"app"
	"log/slog"
	"net/http"
	"strings"

	"github.com/google/uuid"
	"github.com/labstack/echo/v5"
)

func PostHandler(state *app.AppState) echo.HandlerFunc {
	return func(c *echo.Context) error {
		body, err := ReadBody(c)
		if err != nil {
			return err
		}

		task, err := ParseTask(c, body)
		if err != nil {
			return err
		}

		id := uuid.New()

		state.Mu.Lock()
		state.Tasks[id] = task
		state.Mu.Unlock()

		slog.Info(
			"Inserted new task",
			"id", id.String(),
			"secret", len(task.Secret),
			"operation", strings.ToLower(string(task.Operation)),
			"method", c.Request().Method,
		)

		return c.JSON(http.StatusCreated, map[string]string{"id": id.String()})
	}
}
